charmcraft/commands/store/charmlibs.py

```diff
@@ -75,62 +67,69 @@ def create_charm_name_from_importable(charm_name: str) -> str:
+            _api_patch_validator,
+            "Library {!r} metadata field LIBPATCH is not zero or a positive integer.",
+        ),
+        "LIBID": (
+            lambda value: isinstance(value, str) and value and value.isascii(),
+            "Library {!r} metadata field LIBID must be a non-empty ASCII string.",
+        ),
+    }
+
+    # walk the AST "first layer", find assignments to the key fields and validate them
+    metadata = {}
+    for node in ast.iter_child_nodes(tree):
+        if isinstance(node, ast.Assign):
+            for target in node.targets:
+                if target.id in FIELDS:
+                    validator, error_msg = FIELDS[target.id]
+                    if not isinstance(node.value, ast.Constant):
+                        raise CraftError(error_msg.format(str(lib_path)))
+                    real_value = node.value.value
+                    if not validator(real_value):
+                        raise CraftError(error_msg.format(str(lib_path)))
+                    metadata[target.id] = real_value
+
+    # extra verifications for cases that need to consider more than the individual fields
+    missing = set(FIELDS) - set(metadata)
     if missing:
+        joined = ", ".join(sorted(missing))
         raise CraftError(
-            "Library {!r} is missing the mandatory metadata fields: {}.".format(
-                str(lib_path), ", ".join(sorted(missing))
-            )
+            f"Library {str(lib_path)!r} is missing the mandatory metadata fields: {joined}."
         )
-
-    bad_api_patch_msg = "Library {!r} metadata field {} is not zero or a positive integer."
-    try:
-        libapi = get_positive_int(metadata[b"LIBAPI"])
-    except ValueError:
-        raise CraftError(bad_api_patch_msg.format(str(lib_path), "LIBAPI"))
-    try:
-        libpatch = get_positive_int(metadata[b"LIBPATCH"])
-    except ValueError:
-        raise CraftError(bad_api_patch_msg.format(str(lib_path), "LIBPATCH"))
-
-    if libapi == 0 and libpatch == 0:
+    if metadata["LIBAPI"] == 0 and metadata["LIBPATCH"] == 0:
         raise CraftError(
-            "Library {!r} metadata fields LIBAPI and LIBPATCH cannot both be zero.".format(
-                str(lib_path)
-            )
+            f"Library {str(lib_path)!r} metadata fields LIBAPI and LIBPATCH cannot both be zero."
         )
 
-    bad_libid_msg = "Library {!r} metadata field LIBID must be a non-empty ASCII string."
-    try:
-        libid = ast.literal_eval(metadata[b"LIBID"].decode("ascii"))
-    except (ValueError, UnicodeDecodeError):
-        raise CraftError(bad_libid_msg.format(str(lib_path)))
-    if not libid or not isinstance(libid, str):
-        raise CraftError(bad_libid_msg.format(str(lib_path)))
-
-    content = lib_path.read_text()
+    # hash the file excluding those lines where appear keys used for version control
+    metadata_vcs_fields = (b"LIBAPI", b"LIBPATCH", b"LIBID")
+    hasher = hashlib.sha256()
+    with lib_path.open("rb") as fh:
```

Comment:
Pretty sure the Windows test failure is occurring because of this. I *think* forcing `newline="\n"` should do the trick.

Reply:
It was the newline. But `newline` argument does not work in binary mode, so I explicitly replaced/fixed it and added a comment about having the same hash for the same file in all platforms.

Reply:
Thanks!!